File: 1557-minimum-number-of-vertices-to-reach-all-nodes/1557-minimum-number-of-vertices-to-reach-all-nodes.py
class UF:
    def __init__(self):
        self.root = {}
    
    def add(self, a: int) -> None:
        if a not in self.root:
            self.root[a] = a
    
    def find(self, a: int) -> int:
        self.add(a)
        if self.root[a] != a:
            self.root[a] = self.find(self.root[a])
        return self.root[a]

    def union(self, a: int, b: int) -> None:
        A = self.find(a)
        B = self.find(b)
        if A != B:
            self.root[b] = A
    
    def get_roots(self) -> List[int]:
        res = []
        for key, val in self.root.items():
            if int(key) == val:
                res.append(val)
        return res

class Solution:
    def findSmallestSetOfVertices(self, n: int, edges: List[List[int]]) -> List[int]:
        uf = UF()
        for u, v in edges:
            uf.union(u, v)
        return uf.get_roots()

File: 1557-minimum-number-of-vertices-to-reach-all-nodes/1557-minimum-number-of-vertices-to-reach-all-nodes.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var findSmallestSetOfVertices = function(n, edges) {
    const nodes = new Set(Array(n).fill(0).map((_,i) => i))
    for (const [, v] of edges) {
        nodes.delete(v)
    }
    return Array.from(nodes)
};

findSmallestSetOfVertices = function(n, edges) {
    class UF {
        constructor() {
            this.root = {}
        }
        add(n) {
            if (!(n in this.root)) {
                this.root[n] = n
            }
        }
        find(n) {
            this.add(n)
            if (this.root[n] !== n) {
                this.root[n] = this.find(this.root[n])
            }
            return this.root[n]
        }
        union(a, b) {
            const A = this.find(a)
            const B = this.find(b)
            if (A !== B) {
                this.root[b] = A
            }
        }
        getRoots() {
            const res = []
            for (const node in this.root) {
                if (this.root[node] == node) {
                    res.push(this.root[node])
                }
            }
            return res
        }
    }
    const uf = new UF()
    for (const [u, v] of edges) {
        uf.union(u, v)
    }
    return uf.getRoots()
}